designs/plugins-phase-2.md

```diff
@@ -0,0 +1,105 @@
+
+* mdbook-like structure
+    * A Plugin receives a JSON blob plugin "universe" from upstream plugins containing file "nodes", each of which contain file metadata.
+    * The plugin can add or delete nodes (nodes are not mutable), then returns the modified universe so downstream plugins can do the same.
+* Pre-generate a project file for handling complex plugin use-cases.
+* Enhanced "layout" PROJECT file key.
+    * Either an ordered list of plugin {name}/{version}, or ordered map of {name}/{version} to plugin-specific config.
+    * Order matters for certain plugins, ex. a bazel plugin would need to run after all Go files are generated.
+
+## Evaluation of Plugin libraries 
+
+##### Go built-in plugin package vs Hashicorp plugin library vs custom plugin library
+
+Based on the evaluation of plugin libraries such as the built-in go-plugin library and Hashicorp’s plugin library, we have come to the conclusion that it is more suitable to write our own custom typed plugin library.
+
+The built-in plugin library seems to be a non-starter as it is more suitable for in-tree plugins rather than out-of-tree and it doesn’t offer cross language support. Hashicorp’s go plugin system seems more suitable than built-in go-plugin library as it enables cross language/platform support. However, it is more suited for long running plugins as opposed to short lived plugins and it could be overkill with the usage of protobuf.
+ 
+For the stated reasons, the proposal is to write our own plugin system that passes JSON blobs back and forth across stdin/stdout.
+The plugin specification will include type metadata to allow the potential for using other plugin libraries in the future if the need arises. From an implementation standpoint, we could create a new implementation of our plugin interface based on the type specified in the project file. We can do a type switch on the type of plugin the data is being passed to. For ex: if it is native go, then we pass the universe directly & if it is a binary wrapper, then we pass the serialized stream of JSON bytes to it. 
+
+For phase 2 plugin implementation, plugins need a data structure containing a set of file representations to read from and write to that can be passed between multiple plugins. Let's look at that below.
+
+#### Intermediate Representation
+
+The proposed plugin universe is: 
+
+```go
+type File struct {
+    // Path is the file to write
+    Path string `json:"path,omitempty"`
+    
+    // Contents is the generated output
+    Contents string `json:"contents,omitempty"`
+    
+    // Info is the os.FileInfo 
+    Info os.FileInfo  `json:"info,omitempty"`	
+}
+ 
+type Universe struct {
+    // Config stores the project file  
+    Config *config.Config `json:"config,omitempty"`
+
+    // Files contains the model of the files that are being scaffolded
+    Files map[string]File `json:"files,omitempty"`	
```

Comment:
I think it is talking about already templated files. Is there anyway that allows us to pass it before templating and templating after all plugins have run? I think we may be able to abstract part of the templating and scaffolding code so that plugins are easier to develop.